lib/index.js

```diff
@@ -43,46 +43,118 @@ class Flooper{
     this.curBlock = 0;
     this.order = this.flooperItemCount + 1;
     this.speed = speed;
+    this.lastKnownSpeed = this.speed;
 
-    // Bindings
-    this.step = this.step.bind(this);
+    /** 
+     * Bindings
+     */
+    // CORE
+    this.init = this.init.bind(this);
     this.checkBlocks = this.checkBlocks.bind(this);
-  }
-
-  // init
-  init(){
-    requestAnimationFrame(this.step);
-  }
-
-  step(timeStamp){
-    this.left -= this.speed;
-    this.checkBlocks();
+    this.step = this.step.bind(this);
+    this.update = this.update.bind(this);
+    // perf / rAF
+    this.requestTick = this.requestTick.bind(this);
+    this.ticking = false;
 
-    this.flooper.style[this.transformProp] = `translateX(${this.left}px)`
+    // ANIMATION / INTERACTION
+    this.play = this.play.bind(this);
+    this.pause = this.pause.bind(this);
+    this.resume = this.resume.bind(this);
+    this.slowMotion = this.slowMotion.bind(this);
 
-    requestAnimationFrame(this.step);
+    // EVENTS
+    this.addEvents = this.addEvents.bind(this);
   }
 
+    // init
+    init(){
+      this.addEvents();
+      this.play()
+    }
+
+  /**
+   * --------------------
+   * CORE
+   * --------------------
+   */
+  // THE MAGIC
   checkBlocks(){
     // cache block
     let block = this.flooperItem[this.curBlock];
     // cache dimensionsObj
-    let blockSize = block.getBoundingClientRect();
+    let blockRect = block.getBoundingClientRect();
+    let blockComputedStyle = window.getComputedStyle(block);
+    // Check if block has margins
+    let blockSpacings = parseInt(blockComputedStyle.marginLeft, 10) + parseInt(blockComputedStyle.marginRight, 10); 
+    let blockComputedWidth = blockRect.width + blockSpacings;
 
-    if(blockSize.x < -(blockSize.width + this.bufferSize) ) {
+    if(blockRect.x < -(blockComputedWidth + this.bufferSize) ) {
       block.style.order=this.order;
       this.order ++;
-      this.left+=blockSize.width;
+      this.left+=blockComputedWidth;
       this.curBlock++;
 
       if(this.curBlock > this.flooperItemCount - 1){
         this.curBlock = 0;
       }
     }
   }
-}
 
 
-// instanciate and initizialize
-var i = new Flooper();
-i.init();
+  /* rAF */
+  update() {
+    // reset the tick so we can
+    // capture the next onScroll
+    this.ticking = false;
+
+    this.checkBlocks();
+    this.flooper.style[this.transformProp] = `translateX(${this.left}px)`;
+
+    requestAnimationFrame(this.step);
+  }
+
+  step() {
+      this.left -= this.speed; //No IE8
+
+      this.requestTick();
+  }
+
+  requestTick() {
```

Comment:
Why is this requestTick necessary?

Reply:
😕  maybe not.. I was in autopilot mode yesterday. The `requestTick`is used to decide if the animate function should be called or not. I use this pattern on [onScroll animations](https://codepen.io/ohsimtabem/pen/bWZRqa) for the `updating()` fn not being called if user is not scrolling. But since this is an automated continuous animation, you're right and it doesn't make sense.

Lesson: you shouldn't code at 3am 📦